native/thrift_utils.c

```diff
@@ -0,0 +1,105 @@
+#include "thrift_utils.h"
+#include "thrift_consts.h"
+
+int
+dataLen(char *data)
+{
+    int i;
+    int count;
+
+    i = 0;
+    count = 0;
+    while (data[i] != '\0') {
+        if (data[i] == '\\' && data[i+1] == 'x') {
+            count++;
+            i = i+4;
+        }
+        else if (data[i] == '\\') {
+            count++;
+            i = i+2;
+        }
+        else {
+            count++;
+            i++;
+        }
+    }
+    return count;
+}
+
+char *
+detect_protocol(char *data, char *fallbackProtocol)
+{
+    char *protocol;
+
+    if (is_compact_protocol(data)) {
+        protocol = malloc(sizeof(char)*(strlen("TCompactProtocol")+1));
+        strcpy(protocol, "TCompactProtocol");
+        return protocol;
+    }
+    if (is_binary_protocol(data)) {
+        protocol = malloc(sizeof(char)*(strlen("TBinaryProtocol")+1));
+        strcpy(protocol, "TBinaryProtocol");
+        return protocol;
+    }
+    if (is_json_protocol(data)) {
+        protocol = malloc(sizeof(char)*(strlen("TJSONProtocol")+1));
+        strcpy(protocol, "TJSONProtocol");
+        return protocol;
+    }
+    if (fallbackProtocol == NULL) {
+        return NULL;
+    }
+    return fallbackProtocol;
+}
+
+int
+is_compact_protocol(char *data)
+{
+    char aux[5];
+    if (strlen(data) < 1) {
+        return -1;
```

Comment:
you should return 0... otherwise this function will evaluation as true (i.e.: -1 evals as true).